src/java/org/apache/cassandra/cql3/selection/SortedRowsBuilder.java

```diff
@@ -0,0 +1,279 @@
+
+        @Override
+        public void add(List<ByteBuffer> row)
+        {
+            if (toSkip-- <= 0 && rows.size() < limit)
+                rows.add(row);
+        }
+
+        @Override
+        public List<List<ByteBuffer>> build()
+        {
+            return rows;
+        }
+    }
+
+    /**
+     * {@link SortedRowsBuilder} that orders the added rows based on a {@link Comparator}.
+     */
+    private static class WithComparator extends WithCompare<WithCompare.DecoratedRow>
+    {
+        private final Comparator<List<ByteBuffer>> comparator;
+
+        private WithComparator(int limit, int offset, Comparator<List<ByteBuffer>> comparator)
+        {
+            super(limit, offset);
+            this.comparator = comparator;
+        }
+
+        @Override
+        DecoratedRow decorate(List<ByteBuffer> row, int id)
+        {
+            return new DecoratedRow(row, id);
+        }
+
+        @Override
+        public int compare(DecoratedRow one, DecoratedRow other)
+        {
+            return comparator.compare(one.row, other.row);
+        }
+    }
+
+    /**
+     * {@link SortedRowsBuilder} that orders the added rows based on a {@link Index.Scorer}.
+     * </p>
+     * The rows are decorated with their scores, so we don't have to recompute them while sorting.
+     */
+    private static class WithScore extends WithCompare<WithScore.ScoredRow>
+    {
+        private final Index.Scorer scorer;
+
+        private WithScore(int limit, int offset, Index.Scorer scorer)
+        {
+            super(limit, offset);
+            this.scorer = scorer;
+        }
+
+        @Override
+        ScoredRow decorate(List<ByteBuffer> row, int id)
+        {
+            return new ScoredRow(row, id, scorer.score(row));
+        }
+
+        @Override
+        public int compare(ScoredRow one, ScoredRow other)
+        {
+            return scorer.reversed()
+                   ? Float.compare(other.score, one.score)
+                   : Float.compare(one.score, other.score);
+        }
+
+        static final class ScoredRow extends DecoratedRow
+        {
+            private final float score;
+
+            ScoredRow(List<ByteBuffer> row, int id, float score)
+            {
+                super(row, id);
+                this.score = score;
+            }
+        }
+    }
+
+    /**
+     * {@link SortedRowsBuilder} that orders rows based on the {@link #compare(DecoratedRow, DecoratedRow)} method.
+     * </p>
+     * It's possible for {@link #compare(DecoratedRow, DecoratedRow)} to produce ties. To deal with these ties, the rows
+     * are decorated with their position in the sequence of calls to {@link #add(List)}, so we can use that identifying
+     * position to solve ties by favoring the row that was inserted first.
+     * </p>
+     * It keeps at most {@code limit + offset} rows in memory.
+     */
+    private static abstract class WithCompare<T extends WithCompare.DecoratedRow> extends SortedRowsBuilder
```

Comment:
Why do we need a separate `WithCompare` and `WithComparator` classes? I must say, initially that was a bit confusing to me - I get the semantic difference, but anyway, they are named so similarly it is is easy to confuse one with another. The also achieve similar thing but in a slightly different way - one delegates comparison to abstract method, another one delegates to the comparator object. So why isn't `WithComparator` alone enough?  We could create a comparator for dealing with scored rows, so `WithScore` would work fine too. 
I'd prefer to deal with fewer classes in the hierarchy if possible.

Reply:
The reason for extracting a common superclass is that the class of the decorated row, the one that we store in the priority queue, is different. The `decorate` abstract method does that decoration, and I'm not sure how we could skip that one. Perhaps passing it as a `BiFunction<List<ByteBuffer>,Integer>` to the constructor, so we have a single generic class? 
For now, I have renamed `WithCompare` to `WithValuesOrder`,  I'm not sure if it makes it clearer.

Reply:
> Perhaps passing it as a `BiFunction<List<ByteBuffer>, Integer>` to the constructor, so we have a single generic class?


I have just pushed that refactor, hoping it looks clearer. wdyt?

Reply:
I like the refactored code more than the original. The separate decorator is nice. 
Less inheritance, more composition.